### ISSUE TITLE
Add proxy support to IP exclusion

### ISSUE THREAD

**Issue by arderyp** [iipc/openwayback#260]
We have implemented the following IP exclusion on one of our Openwayback 2.1.0 access points:

``` xml
    <property name="requestAuth" value="false" />
    <property name="authentication">
      <bean class="org.archive.wayback.authenticationcontrol.IPMatchesBooleanOperator">
        <property name="allowedRanges">
          <list>
            <value>(IP RANGE)</value>
          </list>
        </property>
      </bean>
    </property>
```

When we access the Openwayback instance directly (via subdomain.loc.gov), the IP exclusion works.  However, we would like to route traffic to the Openwayback instance from a public webserver running on a different VM.  We are achieving this via the following apache rules in the appropriate VirtualHost block in the public (remote) webserver's httpd.conf:

```
RewriteRule ^/somepath/(.*)$ http://subdomain.loc.gov/$1 [P,L]
ProxyPassReverse /somepath/ http://subdomain.loc.gov/
```

The traffic is routed as we expect, but this breaks the IP exclusion.  Assuming our web server's IP is 1.2.3.4 and we are making a request to the public webserver through a machine with IP 5.6.7.8, Catalina.out shows:

```
SEVERE: Unable to parse remote IP address(5.6.7.8, 1.2.3.4)
May 26, 2015 2:21:20 PM org.archive.wayback.webapp.AccessPoint logError
```

I think that, if we make the same request directly, instead of requesting via the proxy webserver, the REQUEST_REMOTE_ADDRESS variable on the WaybackRequest object would be "5.6.7.8".  However, when we proxy the request, the error above suggests that the REQUEST_REMOTE_ADDRESS variable is set to "5.6.7.8, 1.2.3.4".

If this is indeed the case, I would think that a simple check could be added to the getRemoteIPAddress() method of the WaybackRequest object class to check if REQUEST_REMOTE_ADDRESS is a single IP (direct connection) or if it is a comma separated list of IPs (indicating proxy access).   I am not all that familiar with java, but maybe [the following](https://github.com/iipc/openwayback/blob/master/wayback-core/src/main/java/org/archive/wayback/util/IPRange.java#L182-206) could be updated to check for proxy requests before processing the IP:

``` java
         public static byte[] matchIP(String ip) {
                 /** Check if Proxy request*/
                 ip_addresses = ip.split(", ");
                 if (ip_addresses.size() == 2) {
                         ip = ip_addresses[1];
                 } 

                 Matcher m = IP_PATTERN.matcher(ip);
                 if(m != null) {
                         if(m.matches()) {
                                 try {
                                         byte[] res = new byte[4];
                                         for(int i=0; i < 4; i++) {
                                                 int testInt = Integer.parseInt(m.group(i+1));
                                                 if(testInt < 0) {
                                                         return null;
                                                 }
                                                 if(testInt > 255) {
                                                         return null;
                                                 }
                                                 res[i] = (byte) testInt;
                                         }
                                         return res;
                                 } catch(NumberFormatException e) {
                                         e.printStackTrace();
                                         return null;
                                 }
                         }
                 }
                 return null;
         }
```

Or, maybe it would make more sense to actually validate the WaybackRequest.REQUEST_REMOTE_ADDRESS variable before setting it, and leave the IPRange.matchIP() method noted above untouched.  That way, the proper IP is established from the beginning, and just flows downhill into the other functionality of the application.  If this seems like the wiser approach, maybe the following check could be added to the [WaybackRequest.extractHttpRequestInfo()](https://github.com/iipc/openwayback/blob/master/wayback-core/src/main/java/org/archive/wayback/core/WaybackRequest.java#L1110-1118) method:

``` java
         public void extractHttpRequestInfo(HttpServletRequest httpRequest) {

                 putUnlessNull(REQUEST_REFERER_URL, httpRequest.getHeader("REFERER"));

                 String remoteAddr = httpRequest.getHeader("X-Forwarded-For");
                 if (remoteAddr == null) {
                         remoteAddr = httpRequest.getRemoteAddr();
                 }
                 if (remoteAddr != null) {
                     /** Check if Proxy request*/
                     ip_addresses = remoteAddr.split(", ");
                     if (ip_addresses.size() == 2) {
                             put(REQUEST_REMOTE_ADDRESS,ip_addresses[1]);
                             put(REQUEST_REMOTE_PROXY, ip_addresses[0]);
                     } else {
                             put(REQUEST_REMOTE_ADDRESS, remoteAddr)
                     }
                 }
```

If the proxy IP should be stored for some reason, a REQUEST_REMOTE_PROXY string could be added to the class and defined as shown above.  Or, if that info wouldn't be useful in any way, the `put(REQUEST_REMOTE_PROXY, ip_addresses[0]);` line above could be removed.

On the other hand, if you think this could be resolved on our end via some Apache setting, please let me know.  I am looking into it, but not coming up with much so far.

**Comment by kris-sigur** [comment]
Just want to note that the [standard format for the `X-Forwarded-For` header](https://en.wikipedia.org/wiki/X-Forwarded-For#Format) is:

    X-Forwarded-For: client, proxy1, proxy2

So clearly OpenWayback isn't handling this correctly in [extractHttpRequestInfo()](https://github.com/iipc/openwayback/blob/d154bb294991386cc3578c3d5ceb93aed01955fe/wayback-core/src/main/java/org/archive/wayback/core/WaybackRequest.java#L1114-L1118)

**Comment by pardery** [comment]
Thanks for the info Chris

**Comment by ptrourke** [comment]
Because there are tools that are designed to allow users to inject spoofed `X-Forwarded-For` headers into requests ([Firefox X-Forwarded-For Header Addon](https://addons.mozilla.org/en-Us/firefox/addon/x-forwarded-for-header/)), just checking the the leftmost value, the putative `client` value, in the `X-Forwarded-For` header, won't really enforce an IP whitelist. 

> "Since it is easy to forge an X-Forwarded-For field the given information should be used with care. The last IP address is always the IP address that connects to the last proxy, which means it is the most reliable source of information. X-Forwarded-For data can be used in a forward or reverse proxy scenario." [Wikipedia: X-Forwarded-For](https://en.wikipedia.org/wiki/X-Forwarded-For)

The process has to start at the rightmost value in the `X-Forwarded-For` header, iterate leftward through the values, checking that each proxy value is a known proxy, and then check the first value it finds that is not a known proxy against the whitelist.

> When multiple, comma delimited useragent IP addresses are listed in the header value, they are processed in Right-to-Left order. Processing halts when a given useragent IP address is not trusted to present the preceding IP address. The header field is updated to this remaining list of unconfirmed IP addresses, or if all IP addresses were trusted, this header is removed from the request altogether. [Apache Module mod_remoteip](http://httpd.apache.org/docs/2.4/mod/mod_remoteip.html), cited by [Yoav Aner, "Rails IP Spoofing Vulnerabilities and Protection", *GingerLime*](http://blog.gingerlime.com/2012/rails-ip-spoofing-vulnerabilities-and-protection/)

So to do this, I'd suggest adding a configuration option for a PROXY_IP_LIST, then a method or just a block that extracts the real client IP. 

Time permitting, I'm going to take a look at the codebase and see if I can get up to speed with it.

**Comment by johnerikhalse** [comment]
@ptrourke Have you had the time to look into this? The solution you suggest sounds good to me. Can I assign this issue to you?

**Comment by ptrourke** [comment]
Yes, I'll take it.

**Comment by kris-sigur** [comment]
@ptrourke You'd said you had a fix almost ready on the call in November. Is it ready? If not, we may have to defer this to a later release as we'd really like to mint a 2.3.0 release.

**Comment by ptrourke** [comment]
Yes, sorry, I haven't been able to close the deal on this, due to local
difficulties. My apologies.

On Thu, Jan 7, 2016 at 9:23 AM, Kristinn Sigurðsson <
notifications@github.com> wrote:

> @ptrourke <https://github.com/ptrourke> You'd said you had a fix almost
> ready on the call in November. Is it ready? If not, we may have to defer
> this to a later release as we'd really like to mint a 2.3.0 release.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/iipc/openwayback/issues/260#issuecomment-169677867>.
>



-- 
Patrick Rourke
ptrourke [(at)] methymna.com

**Comment by chasehd** [comment]
Hello, all, my name is Chase Dooley, and I'm the newest member of the Web Archiving team at LoC. I'm just getting up to speed, and as such, I started reviewing this ticket. I don't know how far along @ptrourke is with his fix, so this may not be worth noting, but I stumbled across this Tomcat valve: 

https://tomcat.apache.org/tomcat-6.0-doc/api/org/apache/catalina/valves/RemoteIpValve.html

that seems to do what @ptrourke suggested.

```
This valve proceeds as follows:

If the incoming request.getRemoteAddr() matches the valve's list of internal proxies :

- Loop on the comma delimited list of IPs and hostnames passed by the preceding 
  load balancer or proxy in the given request's Http header named $remoteIpHeader 
  (default value x-forwarded-for). Values are processed in right-to-left order.

- For each ip/host of the list:
        - if it matches the internal proxies list, the ip/host is swallowed
        - if it matches the trusted proxies list, the ip/host is added 
           to the created proxies header
        - otherwise, the ip/host is declared to be the remote ip and looping is stopped.

- If the request http header named $protocolHeader (e.g. x-forwarded-for) 
equals to the value of protocolHeaderHttpsValue configuration parameter (default https) 
then request.isSecure = true, request.scheme = https and request.serverPort = 443. 
Note that 443 can be overwritten with the $httpsServerPort configuration parameter.

```

I don't know much about Apache, but this may be a fix. It would put the onus on the server and not the application, however, which seems to be the opposite of the direction this issue was being taken. 

Additionally, I don't know if this would solve the issue .arderyp and @kris-sigur brought up regarding the reversal of the x-forwarded-for in openwayback.

**Comment by chasehd** [comment]
Just for reference, I believe why this worked for LoC in wayback 1.6 and doesn't in Openwayback because of the `extractHttpRequestInfo` method in `WaybackRequest`.

In 1.6, Wayback just added the remote address from `getRemoteAddr()`:

```java

putUnlessNull(REQUEST_REMOTE_ADDRESS, httpRequest.getRemoteAddr());

```

Starting with Openwayback, it checked for the X-Forwarded-For first:

```java

String remoteAddr = httpRequest.getHeader("X-Forwarded-For");
		if (remoteAddr == null) {
			remoteAddr = httpRequest.getRemoteAddr();
		}
putUnlessNull(REQUEST_REMOTE_ADDRESS, remoteAddr);

```
https://github.com/iipc/openwayback/blob/master/wayback-core/src/main/java/org/archive/wayback/core/WaybackRequest.java#L1110

So in 1.6 < the `remoteAddr` was always just one IP and starting with Openwayback it had the possibility to become an "array", which is not what `IPMatchesBooleanOperator` is expecting--this is where I think it gets hung up every time.
https://github.com/iipc/openwayback/blob/829693b6d43f40b1b045f08611a4fa5e27395e29/wayback-core/src/main/java/org/archive/wayback/authenticationcontrol/IPMatchesBooleanOperator.java#L70

**Comment by chasehd** [comment]
So @ptrourke and I have a fix for this. To which branch should I submit the PR, @ldko ?

**Comment by ldko** [comment]
Hi @chasehd, good to hear from you! Please submit to master. Also, if not too much trouble, please update the 2.4.0 release notes section in your PR.

**Comment by ldko** [comment]
fixed via #349
